supervisor/utils/systemd_journal.py

```diff
@@ -0,0 +1,117 @@
+    ts = ts[: ts.index(".") + 4]  # strip TZ offset
+
+    identifier = (
+        f"{entries["SYSLOG_IDENTIFIER"]}[{entries["_PID"]}]"
+        if "_PID" in entries
+        else entries["SYSLOG_IDENTIFIER"]
+    )
+
+    return f"{ts} {entries["_HOSTNAME"]} {identifier}: " f"{entries["MESSAGE"]}"
+
+
+async def journal_logs_reader(
+    journal_logs: ClientResponse,
+    log_formatter: LogFormatter = LogFormatter.PLAIN,
+) -> AsyncGenerator[str, None]:
+    """Read logs from systemd journal line by line, formatted using the given formatter."""
+    match log_formatter:
+        case LogFormatter.PLAIN:
+            formatter_ = journal_plain_formatter
+        case LogFormatter.VERBOSE:
+            formatter_ = journal_verbose_formatter
+        case _:
+            raise ValueError(f"Unknown log format: {log_formatter}")
+
+    async with journal_logs as resp:
+        entries: dict[str, str] = {}
+        while not resp.content.at_eof():
+            line = await resp.content.readuntil(b"\n")
+            # newline means end of message:
+            if line == b"\n":
+                yield formatter_(entries)
+                entries = {}
+                continue
+
+            # Journal fields consisting only of valid non-control UTF-8 codepoints
+            # are serialized as they are (i.e. the field name, followed by '=',
+            # followed by field data), followed by a newline as separator to the next
+            # field. Note that fields containing newlines cannot be formatted like
+            # this. Non-control UTF-8 codepoints are the codepoints with value at or
+            # above 32 (' '), or equal to 9 (TAB).
+            idx = line.find(b"=")
+            if idx != -1:
+                # name=value\n
+                name = line[:idx].decode("utf-8")
+                if name not in formatter_.required_fields:
+                    continue
+                data = line[idx + 1 : -1]  # strip \n
+            else:
+                # Other journal fields are serialized in a special binary safe way:
+                # field name, followed by newline
+                name = line[:-1].decode("utf-8")  # strip \n
+                # followed by a binary 64-bit little endian size value,
+                length_raw = await resp.content.readexactly(8)
+                length = int.from_bytes(length_raw, byteorder="little")
+                # followed by the binary field data,
+                data = await resp.content.readexactly(length)
+                # followed by a newline as separator to the next field.
+                if (await resp.content.readexactly(1)) != b"\n":
+                    raise MalformedBinaryEntry(line)
+                if name not in formatter_.required_fields:
+                    # we must read to the end of the entry in the stream, so we can
+                    # only continue the loop here
+                    continue
+
+            entries[name] = data.decode("utf-8")
```

Comment:
We can combine the continue like so:

```suggestion
            name, sep, data = line.partition(b'=')
            if not sep:
                # Other journal fields are serialized in a special binary safe way:
                # field name, followed by newline
                name = name[:-1]  # strip \n
                # followed by a binary 64-bit little endian size value,
                length_raw = await resp.content.readexactly(8)
                length = int.from_bytes(length_raw, byteorder="little")
                # followed by the binary field data,
                data = await resp.content.readexactly(length)
                # followed by a newline as separator to the next field.
                if (await resp.content.readexactly(1)) != b"\n":
                    raise MalformedBinaryEntry(line)

            name = name.decode("utf-8")
            if name not in formatter_.required_fields:
                # we must read to the end of the entry in the stream, so we can
                # only continue the loop here
                continue
            entries[name] = data.decode("utf-8")
```

Reply:
Committed with minor adjustments.